File: wayback-downloader/wayback_downloader/downloader.py
"""Threaded downloader that fetches archived snapshots to disk and reports
progress through a thread-safe queue of events."""

import os
import queue
import threading
from concurrent.futures import ThreadPoolExecutor
from typing import List, Optional

import requests

from .cdx import Snapshot
from .urlmap import url_to_filepath

USER_AGENT = "wayback-downloader-web/1.0 (+https://github.com/hartator/wayback-machine-downloader)"


class DownloadJob:
    """Downloads a list of snapshots to `output_dir`, emitting progress
    events onto `self.events` (a queue.Queue) as it goes."""

    def __init__(
        self,
        snapshots: List[Snapshot],
        output_dir: str,
        *,
        concurrency: int = 8,
        all_snapshots: bool = False,
        timeout: int = 30,
    ):
        self.snapshots = snapshots
        self.output_dir = output_dir
        self.concurrency = max(1, min(int(concurrency or 8), 20))
        self.all_snapshots = all_snapshots
        self.timeout = timeout

        self.total = len(snapshots)
        self.completed = 0
        self.failed = 0
        self.skipped = 0

        self.status = "pending"  # pending -> running -> done|cancelled|error
        self.error: Optional[str] = None

        self.events: "queue.Queue[dict]" = queue.Queue()
        self._cancel_event = threading.Event()
        self._lock = threading.Lock()

    def cancel(self):
        self._cancel_event.set()

    def snapshot_dict(self) -> dict:
        with self._lock:
            return {
                "status": self.status,
                "total": self.total,
                "completed": self.completed,
                "failed": self.failed,
                "skipped": self.skipped,
                "error": self.error,
            }

    def _emit(self, event_type: str, **extra):
        payload = {"type": event_type, **self.snapshot_dict()}
        payload.update(extra)
        self.events.put(payload)

    def _target_path(self, snap: Snapshot) -> str:
        rel = url_to_filepath(snap.original)
        if self.all_snapshots:
            host, _, rest = rel.partition("/")
            rel = f"{host}/{snap.timestamp}/{rest or 'index.html'}"
        # Normalize to the host OS separator and keep everything inside output_dir.
        return os.path.normpath(os.path.join(self.output_dir, *rel.split("/")))

    def _download_one(self, session: requests.Session, snap: Snapshot):
        if self._cancel_event.is_set():
            return

        target = self._target_path(snap)
        os.makedirs(os.path.dirname(target), exist_ok=True)

        if os.path.exists(target):
            with self._lock:
                self.skipped += 1
            self._emit("skip", url=snap.original)
            return

        try:
            resp = session.get(snap.archive_url, timeout=self.timeout)
            resp.raise_for_status()
            with open(target, "wb") as fh:
                fh.write(resp.content)
            with self._lock:
                self.completed += 1
            self._emit("ok", url=snap.original, size=len(resp.content))
        except Exception as exc:  # noqa: BLE001 - report every failure, keep going
            with self._lock:
                self.failed += 1
            self._emit("error", url=snap.original, message=str(exc))

    def run(self):
        self.status = "running"
        self._emit("status")

        if not self.snapshots:
            self.status = "done"
            self._emit("status")
            return

        os.makedirs(self.output_dir, exist_ok=True)
        session = requests.Session()
        session.headers["User-Agent"] = USER_AGENT

        try:
            with ThreadPoolExecutor(max_workers=self.concurrency) as pool:
                futures = [pool.submit(self._download_one, session, s) for s in self.snapshots]
                for fut in futures:
                    fut.result()
                    if self._cancel_event.is_set():
                        break
        except Exception as exc:  # noqa: BLE001
            self.status = "error"
            self.error = str(exc)
            self._emit("status")
            return
        finally:
            session.close()

        if self._cancel_event.is_set():
            self.status = "cancelled"
        else:
            self.status = "done"
        self._emit("status")

File: wayback-downloader/src/lib/downloader.ts
import { archiveUrl, type Snapshot } from './cdx';
import { urlToFilepath } from './urlmap';

export interface DownloadedFile {
    name: string;
    blob: Blob;
    size: number;
}

export interface DownloadProgress {
    completed: number;
    failed: number;
    skipped: number;
    total: number;
}

export interface DownloadEvent {
    type: 'ok' | 'error' | 'skip';
    url: string;
    message?: string;
}

export interface DownloadOptions {
    concurrency?: number;
    allSnapshots?: boolean;
    timeoutMs?: number;
    onEvent?: (event: DownloadEvent, progress: DownloadProgress) => void;
    isCancelled?: () => boolean;
}

export interface DownloadResult {
    files: DownloadedFile[];
    completed: number;
    failed: number;
    skipped: number;
    cancelled: boolean;
}

function targetPath(snap: Snapshot, allSnapshots: boolean): string {
    const relPath = urlToFilepath(snap.original);
    if (!allSnapshots) return relPath;
    const [host, ...rest] = relPath.split('/');
    return `${host}/${snap.timestamp}/${rest.join('/') || 'index.html'}`;
}

/** Downloads every snapshot directly from the browser (not through a
 * server), so requests go out on the visitor's own network rather than a
 * datacenter IP range that archive.org may throttle or drop. */
export async function downloadSnapshots(snapshots: Snapshot[], opts: DownloadOptions = {}): Promise<DownloadResult> {
    const concurrency = Math.max(1, Math.min(opts.concurrency ?? 6, 12));
    const timeoutMs = opts.timeoutMs ?? 15000;
    const total = snapshots.length;

    const files: DownloadedFile[] = [];
    const seen = new Set<string>();
    let completed = 0;
    let failed = 0;
    let skipped = 0;

    const emit = (event: DownloadEvent) => {
        opts.onEvent?.(event, { completed, failed, skipped, total });
    };

    async function downloadOne(snap: Snapshot): Promise<void> {
        const relPath = targetPath(snap, Boolean(opts.allSnapshots));

        if (seen.has(relPath)) {
            skipped += 1;
            emit({ type: 'skip', url: snap.original });
            return;
        }
        seen.add(relPath);

        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), timeoutMs);
        try {
            const resp = await fetch(archiveUrl(snap), { signal: controller.signal });
            if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
            const blob = await resp.blob();
            files.push({ name: relPath, blob, size: blob.size });
            completed += 1;
            emit({ type: 'ok', url: snap.original });
        } catch (err) {
            failed += 1;
            emit({ type: 'error', url: snap.original, message: (err as Error).message });
        } finally {
            clearTimeout(timer);
        }
    }

    let next = 0;
    let cancelled = false;
    async function worker(): Promise<void> {
        while (next < snapshots.length) {
            if (opts.isCancelled?.()) {
                cancelled = true;
                return;
            }
            const snap = snapshots[next++];
            await downloadOne(snap);
        }
    }

    await Promise.all(Array.from({ length: Math.min(concurrency, snapshots.length) }, worker));

    return { files, completed, failed, skipped, cancelled };
}
